src/Ilios/CoreBundle/Classes/CourseRollover.php

```diff
@@ -0,0 +1,399 @@
+        $newStartDate = (!empty($options['new-start-date'])) ? new \DateTime($options['new-start-date']) : null;
+
+        //make sure that the new course's academic year or new start date year is not in the past
+        $this->confirmYearIsNotInPast($newAcademicYear);
+        if (!empty($newStartDate)) {
+            $this->confirmYearIsNotInPast($newStartDate->format('Y'));
+        }
+
+        //get the original course object
+        $originalCourse = $this->getOriginalCourse($originalCourseId);
+
+        //if a new title is to be used, update before checking for duplicates
+        $newTitle = !empty($options['new-course-title']) ? $options['new-course-title'] : $originalCourse->getTitle();
+        //before creating the newCourse object, check for courses with same title & year, so a rollover is not run 2x
+        $this->checkForDuplicateRollover($newTitle, $newAcademicYear);
+
+        //get/set the original course's start/end dates for use in calculation of offset
+        $originalCourseStartDate = $originalCourse->getStartDate();
+        $originalCourseEndDate = $originalCourse->getEndDate();
+
+        //get/set the offset in weeks and its +/- sign and set it as globally-accessible string in the class
+        $weeksOffset = $this->calculateRolloverOffsetInWeeks($originalCourse, $newAcademicYear, $newStartDate);
+        $weeksOffsetModifier = gmp_sign($weeksOffset);
+        $offsetInWeeks = (($weeksOffsetModifier < 0) ? '-' : '+') . ' ' . abs($weeksOffset) . ' weeks';
+
+        //create/modify the newCourse start and end dates based on the original dates and the offset in weeks
+        $newCourseStartDate = clone $originalCourseStartDate;
+        $newCourseStartDate->modify($offsetInWeeks);
+        $newCourseEndDate = clone $originalCourseEndDate;
+        $newCourseEndDate->modify($offsetInWeeks);
+
+        //create the Course
+        //if there are not any duplicates, create a new course with the relevant info
+        /* @var CourseInterface $newCourse */
+        $newCourse = $this->courseManager->create();
+        $newCourse->setTitle($newTitle);
+        $newCourse->setLevel($originalCourse->getLevel());
+        $newCourse->setYear($newAcademicYear);
+        $newCourse->setStartDate($newCourseStartDate);
+        $newCourse->setEndDate($newCourseEndDate);
+        $newCourse->setExternalId($originalCourse->getExternalId());
+        $newCourse->setLocked(0);
+        $newCourse->setArchived(0);
+        $newCourse->setPublishedAsTbd(0);
+        $newCourse->setPublished(0);
+        $newCourse->setClerkshipType($originalCourse->getClerkshipType());
+        $newCourse->setSchool($originalCourse->getSchool());
+        $newCourse->setDirectors($originalCourse->getDirectors());
+        $newCourse->setTerms($originalCourse->getTerms());
+        $newCourse->setObjectives($originalCourse->getObjectives());
+        $newCourse->setMeshDescriptors($originalCourse->getMeshDescriptors());
```

Comment:
correct, needs to be wrapped with a conditional that checks `skip-course-mesh` option.